// gzipper/src/Config.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Config = void 0;
var path_1 = __importDefault(require("path"));
var util_1 = __importDefault(require("util"));
var fs_1 = __importDefault(require("fs"));
var constants_1 = require("./constants");
var helpers_1 = require("./helpers");
var Config = /** @class */ (function () {
    /**
     * Creates an instance of Config.
     */
    function Config() {
        this.nativeFs = {
            writeFile: util_1.default.promisify(fs_1.default.writeFile),
            exists: util_1.default.promisify(fs_1.default.exists),
        };
        this._configContent = {};
        this._configFile = path_1.default.resolve(process.cwd(), constants_1.CONFIG_FOLDER, constants_1.CONFIG_FILE);
        this.setProperty('version', helpers_1.Helpers.getVersion());
    }
    Object.defineProperty(Config.prototype, "configContent", {
        get: function () {
            return this._configContent;
        },
        set: function (value) {
            this._configContent = value;
        },
        enumerable: false,
        configurable: true
    });
    /**
     * Read config (.gzipperconfig).
     */
    Config.prototype.readConfig = function () {
        return __awaiter(this, void 0, void 0, function () {
            var response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.nativeFs.exists(this._configFile)];
                    case 1:
                        if (!_a.sent()) return [3 /*break*/, 3];
                        return [4 /*yield*/, helpers_1.Helpers.readFile(this._configFile)];
                    case 2:
                        response = _a.sent();
                        this._configContent = JSON.parse(response.toString());
                        _a.label = 3;
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * set additional data for property to config file (.gzipperconfig).
     */
    Config.prototype.setProperty = function (field, content) {
        this._configContent[field] = content;
    };
    /**
     * delete property from config file (.gzipperconfig).
     */
    Config.prototype.deleteProperty = function (field) {
        delete this._configContent[field];
    };
    /**
     * Init or update config (.gzipperconfig).
     */
    Config.prototype.writeConfig = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.nativeFs.writeFile(path_1.default.resolve(this._configFile), JSON.stringify(this._configContent, null, 2))];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    return Config;
}());
exports.Config = Config;
