// gzipper/src/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.COMPRESSION_EXTENSIONS = exports.CONFIG_FILE = exports.CACHE_FOLDER = exports.CONFIG_FOLDER = exports.DEFAULT_OUTPUT_FORMAT_MESSAGE = exports.INCREMENTAL_ENABLE_MESSAGE = exports.NO_PATH_MESSAGE = exports.WORKER_STARTED = exports.NO_FILES_MESSAGE = exports.OUTPUT_FILE_FORMAT_REGEXP = void 0;
var enums_1 = require("./enums");
exports.OUTPUT_FILE_FORMAT_REGEXP = /(\[filename\]*)|(\[hash\]*)|(\[compressExt\]*)|(\[ext\]*)/g;
exports.NO_FILES_MESSAGE = 'No files for compression.';
exports.WORKER_STARTED = 'Worker has started.';
exports.NO_PATH_MESSAGE = "Can't find a path.";
exports.INCREMENTAL_ENABLE_MESSAGE = 'Incremental compression has been enabled.';
exports.DEFAULT_OUTPUT_FORMAT_MESSAGE = 'Default output file format: [filename].[ext].[compressExt]';
exports.CONFIG_FOLDER = '.gzipper';
exports.CACHE_FOLDER = 'cache';
exports.CONFIG_FILE = '.gzipperconfig';
exports.COMPRESSION_EXTENSIONS = [
    enums_1.CompressionExtensions.GZIP,
    enums_1.CompressionExtensions.DEFLATE,
    enums_1.CompressionExtensions.BROTLI,
];
