// gzipper/src/Incremental.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Incremental = void 0;
var fs_1 = __importDefault(require("fs"));
var crypto_1 = __importDefault(require("crypto"));
var path_1 = __importDefault(require("path"));
var util_1 = __importDefault(require("util"));
var uuid_1 = require("uuid");
var deep_equal_1 = __importDefault(require("deep-equal"));
var constants_1 = require("./constants");
var helpers_1 = require("./helpers");
var Incremental = /** @class */ (function () {
    /**
     * Creates an instance of Incremental.
     */
    function Incremental(config) {
        this.nativeFs = {
            exists: util_1.default.promisify(fs_1.default.exists),
            unlink: util_1.default.promisify(fs_1.default.unlink),
            readdir: util_1.default.promisify(fs_1.default.readdir),
            lstat: util_1.default.promisify(fs_1.default.lstat),
            rmdir: util_1.default.promisify(fs_1.default.rmdir),
        };
        this._filePaths = new Map();
        if (config) {
            this.config = config;
        }
        this._cacheFolder = path_1.default.resolve(process.cwd(), constants_1.CONFIG_FOLDER, constants_1.CACHE_FOLDER);
    }
    Object.defineProperty(Incremental.prototype, "cacheFolder", {
        get: function () {
            return this._cacheFolder;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Incremental.prototype, "filePaths", {
        get: function () {
            return Object.fromEntries(this._filePaths);
        },
        set: function (value) {
            this._filePaths = new Map(Object.entries(value));
        },
        enumerable: false,
        configurable: true
    });
    /**
     * Read config (.gzipperconfig).
     */
    Incremental.prototype.readConfig = function () {
        return __awaiter(this, void 0, void 0, function () {
            var incrementalConfig;
            return __generator(this, function (_a) {
                incrementalConfig = this.config.configContent.incremental;
                if (incrementalConfig) {
                    this._filePaths = new Map(Object.entries(incrementalConfig.files));
                }
                return [2 /*return*/];
            });
        });
    };
    /**
     * update config (.gzipperconfig).
     */
    Incremental.prototype.updateConfig = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this.config.setProperty('incremental', {
                    files: helpers_1.Helpers.mapToJSON(this._filePaths),
                });
                return [2 /*return*/];
            });
        });
    };
    /**
     * Create cache folder (.gzipper).
     */
    Incremental.prototype.initCacheFolder = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.nativeFs.exists(this._cacheFolder)];
                    case 1:
                        if (!!(_a.sent())) return [3 /*break*/, 3];
                        return [4 /*yield*/, helpers_1.Helpers.createFolders(this._cacheFolder)];
                    case 2:
                        _a.sent();
                        _a.label = 3;
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Returns file incremental info and save checksum and options to `filePath` (if file is changed or newly created).
     */
    Incremental.prototype.setFile = function (target, checksum, compressionType, compressOptions) {
        var filePath = this._filePaths.get(target);
        var selectedRevision = filePath === null || filePath === void 0 ? void 0 : filePath.revisions.find(function (revision) {
            return compressionType === revision.compressionType &&
                (0, deep_equal_1.default)(revision.options, compressOptions);
        });
        if (!filePath) {
            var fileId = (0, uuid_1.v4)();
            this._filePaths.set(target, {
                revisions: [
                    {
                        lastChecksum: checksum,
                        fileId: fileId,
                        date: new Date(),
                        compressionType: compressionType,
                        options: compressOptions,
                    },
                ],
            });
            return {
                isChanged: true,
                fileId: fileId,
            };
        }
        if (!selectedRevision) {
            var fileId = (0, uuid_1.v4)();
            this._filePaths.set(target, {
                revisions: filePath.revisions.concat({
                    lastChecksum: checksum,
                    fileId: fileId,
                    date: new Date(),
                    compressionType: compressionType,
                    options: compressOptions,
                }),
            });
            return {
                isChanged: true,
                fileId: fileId,
            };
        }
        if (selectedRevision.lastChecksum !== checksum) {
            this._filePaths.set(target, {
                revisions: filePath.revisions.map(function (revision) {
                    return revision.fileId === selectedRevision.fileId
                        ? __assign(__assign({}, revision), { date: new Date(), lastChecksum: checksum }) : revision;
                }),
            });
            return {
                isChanged: true,
                fileId: selectedRevision.fileId,
            };
        }
        return {
            isChanged: false,
            fileId: selectedRevision.fileId,
        };
    };
    /**
     * Returns file checksum.
     */
    Incremental.prototype.getFileChecksum = function (target) {
        return __awaiter(this, void 0, void 0, function () {
            var hash, stream;
            return __generator(this, function (_a) {
                hash = crypto_1.default.createHash('md5');
                stream = fs_1.default.createReadStream(target);
                return [2 /*return*/, new Promise(function (resolve, reject) {
                        stream.on('data', function (data) { return hash.update(data, 'utf8'); });
                        stream.on('end', function () { return resolve(hash.digest('hex')); });
                        stream.on('error', function (error) { return reject(error); });
                    })];
            });
        });
    };
    /**
     * Purge cache folder.
     */
    Incremental.prototype.cachePurge = function () {
        return __awaiter(this, void 0, void 0, function () {
            var recursiveRemove;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.nativeFs.exists(this._cacheFolder)];
                    case 1:
                        if (!(_a.sent())) {
                            throw new Error('No cache found.');
                        }
                        recursiveRemove = function (folderPath) {
                            if (folderPath === void 0) { folderPath = _this._cacheFolder; }
                            return __awaiter(_this, void 0, void 0, function () {
                                var files, files_1, files_1_1, file, filePath, isDirectory, e_1_1;
                                var e_1, _a;
                                return __generator(this, function (_b) {
                                    switch (_b.label) {
                                        case 0: return [4 /*yield*/, this.nativeFs.readdir(folderPath)];
                                        case 1:
                                            files = _b.sent();
                                            _b.label = 2;
                                        case 2:
                                            _b.trys.push([2, 10, 11, 12]);
                                            files_1 = __values(files), files_1_1 = files_1.next();
                                            _b.label = 3;
                                        case 3:
                                            if (!!files_1_1.done) return [3 /*break*/, 9];
                                            file = files_1_1.value;
                                            filePath = path_1.default.resolve(folderPath, file);
                                            return [4 /*yield*/, this.nativeFs.lstat(filePath)];
                                        case 4:
                                            isDirectory = (_b.sent()).isDirectory();
                                            if (!isDirectory) return [3 /*break*/, 6];
                                            return [4 /*yield*/, recursiveRemove(filePath)];
                                        case 5:
                                            _b.sent();
                                            return [3 /*break*/, 8];
                                        case 6: return [4 /*yield*/, this.nativeFs.unlink(filePath)];
                                        case 7:
                                            _b.sent();
                                            _b.label = 8;
                                        case 8:
                                            files_1_1 = files_1.next();
                                            return [3 /*break*/, 3];
                                        case 9: return [3 /*break*/, 12];
                                        case 10:
                                            e_1_1 = _b.sent();
                                            e_1 = { error: e_1_1 };
                                            return [3 /*break*/, 12];
                                        case 11:
                                            try {
                                                if (files_1_1 && !files_1_1.done && (_a = files_1.return)) _a.call(files_1);
                                            }
                                            finally { if (e_1) throw e_1.error; }
                                            return [7 /*endfinally*/];
                                        case 12: return [4 /*yield*/, this.nativeFs.rmdir(folderPath)];
                                        case 13:
                                            _b.sent();
                                            return [2 /*return*/];
                                    }
                                });
                            });
                        };
                        return [4 /*yield*/, recursiveRemove()];
                    case 2:
                        _a.sent();
                        this.config.deleteProperty('incremental');
                        return [4 /*yield*/, this.config.writeConfig()];
                    case 3:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Returns cache size.
     */
    Incremental.prototype.cacheSize = function (folderPath, size) {
        if (folderPath === void 0) { folderPath = this._cacheFolder; }
        if (size === void 0) { size = 0; }
        return __awaiter(this, void 0, void 0, function () {
            var files, files_2, files_2_1, file, filePath, fileStat, _a, e_2_1;
            var e_2, _b;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0: return [4 /*yield*/, this.nativeFs.exists(this._cacheFolder)];
                    case 1:
                        if (!(_c.sent())) {
                            throw new Error('No cache found.');
                        }
                        return [4 /*yield*/, this.nativeFs.readdir(folderPath)];
                    case 2:
                        files = _c.sent();
                        if (!files.length) {
                            return [2 /*return*/, 0];
                        }
                        _c.label = 3;
                    case 3:
                        _c.trys.push([3, 10, 11, 12]);
                        files_2 = __values(files), files_2_1 = files_2.next();
                        _c.label = 4;
                    case 4:
                        if (!!files_2_1.done) return [3 /*break*/, 9];
                        file = files_2_1.value;
                        filePath = path_1.default.resolve(folderPath, file);
                        return [4 /*yield*/, this.nativeFs.lstat(filePath)];
                    case 5:
                        fileStat = _c.sent();
                        if (!fileStat.isDirectory()) return [3 /*break*/, 7];
                        _a = size;
                        return [4 /*yield*/, this.cacheSize(filePath, size)];
                    case 6:
                        size = _a + _c.sent();
                        return [3 /*break*/, 8];
                    case 7:
                        if (fileStat.isFile()) {
                            size += fileStat.size;
                        }
                        _c.label = 8;
                    case 8:
                        files_2_1 = files_2.next();
                        return [3 /*break*/, 4];
                    case 9: return [3 /*break*/, 12];
                    case 10:
                        e_2_1 = _c.sent();
                        e_2 = { error: e_2_1 };
                        return [3 /*break*/, 12];
                    case 11:
                        try {
                            if (files_2_1 && !files_2_1.done && (_b = files_2.return)) _b.call(files_2);
                        }
                        finally { if (e_2) throw e_2.error; }
                        return [7 /*endfinally*/];
                    case 12: return [2 /*return*/, size];
                }
            });
        });
    };
    return Incremental;
}());
exports.Incremental = Incremental;
