// gzipper/src/compressions/Brotli.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BrotliCompression = void 0;
var zlib_1 = __importDefault(require("zlib"));
var Compression_1 = require("./Compression");
var enums_1 = require("../enums");
/**
 * Brotli compression
 */
var BrotliCompression = /** @class */ (function (_super) {
    __extends(BrotliCompression, _super);
    /**
     * Creates an instance of BrotliCompression
     */
    function BrotliCompression(options) {
        var _this = _super.call(this, options) || this;
        _this.compressionName = enums_1.CompressionNames.BROTLI;
        _this.ext = enums_1.CompressionExtensions.BROTLI;
        return _this;
    }
    /**
     * Returns brotli compression instance in closure.
     */
    BrotliCompression.prototype.getCompression = function () {
        return zlib_1.default.createBrotliCompress({
            params: this.compressionOptions,
        });
    };
    /**
     * Returns human-readable brotli compression options info.
     */
    BrotliCompression.prototype.readableOptions = function () {
        return _super.prototype.readableOptions.call(this, this.getBrotliOptionName.bind(this));
    };
    /**
     * Build brotli options object [compressionOptions].
     */
    BrotliCompression.prototype.selectCompression = function () {
        var options = {};
        if (this.options.brotliParamMode !== undefined) {
            switch (this.options.brotliParamMode) {
                case 'text':
                    options[zlib_1.default.constants.BROTLI_PARAM_MODE] =
                        zlib_1.default.constants.BROTLI_MODE_TEXT;
                    break;
                case 'font':
                    options[zlib_1.default.constants.BROTLI_PARAM_MODE] =
                        zlib_1.default.constants.BROTLI_MODE_FONT;
                    break;
                case 'default':
                default:
                    options[zlib_1.default.constants.BROTLI_PARAM_MODE] =
                        zlib_1.default.constants.BROTLI_MODE_GENERIC;
                    break;
            }
        }
        if (this.options.brotliQuality !== undefined) {
            options[zlib_1.default.constants.BROTLI_PARAM_QUALITY] = this.options.brotliQuality;
        }
        if (this.options.brotliSizeHint !== undefined) {
            options[zlib_1.default.constants.BROTLI_PARAM_SIZE_HINT] =
                this.options.brotliSizeHint;
        }
        this.compressionOptions = options;
    };
    /**
     * Returns human-readable brotli option name.
     */
    BrotliCompression.prototype.getBrotliOptionName = function (index) {
        switch (parseInt(index)) {
            case zlib_1.default.constants.BROTLI_PARAM_MODE:
                return 'paramMode';
            case zlib_1.default.constants.BROTLI_PARAM_QUALITY:
                return 'quality';
            case zlib_1.default.constants.BROTLI_PARAM_SIZE_HINT:
                return 'sizeHint';
        }
    };
    return BrotliCompression;
}(Compression_1.Compression));
exports.BrotliCompression = BrotliCompression;
